File: solutions/typescript/bank-account/1/bank-account.ts
export class ValueError extends Error {
    constructor() {
        super('Bank account error');
    }
}

export class BankAccount {
    private _balance = 0;
    private isOpen = false;

    public open(): void {
        if (this.isOpen) {
            throw new ValueError();
        }
        this.isOpen = true;
        this._balance = 0;
    }

    public close(): void {
        if (!this.isOpen) {
            throw new ValueError();
        }
        this.isOpen = false;
    }

    public deposit(amount: number): void {
        if (!this.isOpen || amount < 0) {
            throw new ValueError();
        }
        this._balance += amount;
    }

    public withdraw(amount: number): void {
        if (!this.isOpen || amount < 0 || amount > this._balance) {
            throw new ValueError();
        }
        this._balance -= amount;
    }

    public get balance(): number {
        if (!this.isOpen) {
            throw new ValueError();
        }
        return this._balance;
    }
}

File: solutions/lua/bank-account/1/bank-account.lua
local BankAccount = {}

function BankAccount:new()
    return setmetatable({ _balance = 0, _isOpen = false }, { __index = self })
end

function BankAccount:open()
    if self._isOpen then
        error()
    end
    self._balance = 0
    self._isOpen = true
end

function BankAccount:close()
    if not self._isOpen then
        error()
    end
    self._isOpen = false
end

function BankAccount:deposit(amount)
    if not self._isOpen or amount < 0 then
        error()
    end
    self._balance = self._balance + amount
end

function BankAccount:withdraw(amount)
    if not self._isOpen or amount < 0 or amount > self._balance then
        error()
    end
    self._balance = self._balance - amount
end

function BankAccount:balance()
    if not self._isOpen then
        error()
    end
    return self._balance
end


return BankAccount
